builtin/stash--helper.c

```diff
@@ -650,53 +650,67 @@ static int git_stash_config(const char *var, const char *value, void *cb)
 		show_patch = git_config_bool(var, value);
 		return 0;
 	}
+	return git_default_config(var, value, cb);
 }
 
 static int show_stash(int argc, const char **argv, const char *prefix)
 {
 	int i, ret = 0;
-	struct argv_array args = ARGV_ARRAY_INIT;
-	struct argv_array args_refs = ARGV_ARRAY_INIT;
 	struct stash_info info;
-	struct option options[] = {
-		OPT_END()
-	};
+	struct rev_info rev;
+	struct argv_array stash_args = ARGV_ARRAY_INIT;
+	int flags = 0;
 
-	argc = parse_options(argc, argv, prefix, options,
-			git_stash_helper_show_usage, PARSE_OPT_KEEP_UNKNOWN);
+	init_diff_ui_defaults();
+	git_config(git_diff_ui_config, NULL);
 
-	argv_array_push(&args, "diff");
+	init_revisions(&rev, prefix);
 
-	for (i = 0; i < argc; ++i) {
-		if (argv[i][0] == '-')
-			argv_array_push(&args, argv[i]);
+	/* Push args which are not options into stash_args. */
+	for (i = 1; i < argc; ++i) {
+		if (argv[i][0] != '-')
+			argv_array_push(&stash_args, argv[i]);
 		else
-			argv_array_push(&args_refs, argv[i]);
-	}
-
-	if (get_stash_info(&info, args_refs.argc, args_refs.argv)) {
-		argv_array_clear(&args);
-		argv_array_clear(&args_refs);
-		return -1;
+			flags++;
 	}
 
-	if (args.argc == 1) {
+	/*
+	 * The config settings are applied only if there are not passed
+	 * any flags.
+	 */
+	if (!flags) {
 		git_config(git_stash_config, NULL);
 		if (show_stat)
-			argv_array_push(&args, "--stat");
-
-		if (show_patch)
-			argv_array_push(&args, "-p");
+			rev.diffopt.output_format |= DIFF_FORMAT_DIFFSTAT;
+		if (show_patch) {
+			rev.diffopt.output_format = ~DIFF_FORMAT_NO_OUTPUT;
+			rev.diffopt.output_format |= DIFF_FORMAT_PATCH;
+		}
 	}
 
-	argv_array_pushl(&args, oid_to_hex(&info.b_commit),
-		oid_to_hex(&info.w_commit), NULL);
+	if (get_stash_info(&info, stash_args.argc, stash_args.argv))
+		return -1;
 
-	free_stash_info(&info);
+	argc = setup_revisions(argc, argv, &rev, NULL);
+	if (!rev.diffopt.output_format) {
+		rev.diffopt.output_format = DIFF_FORMAT_PATCH;
+	}
+	diff_setup_done(&rev.diffopt);
+	rev.diffopt.flags.recursive = 1;
+	setup_diff_pager(&rev.diffopt);
 
-	ret = cmd_diff(args.argc, args.argv, prefix);
+	/*
+	 * We can early return if there was any option not recognised by
```

Comment:
"We can return early ..."

And we still have to release the memory (`stash_args`) if we return early.